src/pip/_internal/index.py

```diff
@@ -226,6 +226,119 @@ def _get_html_page(link, session=None):
         return HTMLPage(resp.content, resp.url, resp.headers)
 
 
+class _LinkParseError(ValueError):
+    pass
+
+
+class _LinkNotAFile(_LinkParseError):
+    pass
+
+
+class _LinkFormatUnsupported(_LinkParseError):
+    pass
+
+
+class _BinaryLinkNotPermitted(_LinkParseError):
+    pass
+
+
+class _MacOSXZip(_LinkParseError):
+    pass
+
+
+class _IncorrectWheelName(_LinkParseError):
+    pass
+
+
+class _IncompatibleWheel(_LinkParseError):
+    pass
+
+
+class _SourceLinkNotPermitted(_LinkParseError):
+    pass
+
+
+class _PythonVersionNotSupported(_LinkParseError):
+    pass
+
+
+def _egg_info_matches(egg_info, canonical_name):
+    """Pull the version part out of a string.
+
+    :param egg_info: The string to parse. E.g. foo-2.1
+    :param canonical_name: The canonicalized name of the package this
+        belongs to.
+    """
+    try:
+        version_start = _find_name_version_sep(egg_info, canonical_name) + 1
+    except ValueError:
+        return None
+    version = egg_info[version_start:]
+    if not version:
+        return None
+    return version
+
+
+_PYTHON_VERSION_RE = re.compile(r'-py([123]\.?[0-9]?)$')
+
+
+def _parse_package_version_non_wheel(egg_info, search):
+    # This should be up by the search.ok_binary check, but see issue 2700.
+    if "source" not in search.formats:
+        raise _SourceLinkNotPermitted
+
+    info = _egg_info_matches(egg_info, search.canonical)
```

Comment:
I just noticed this, but since you're converting `_egg_info_matches()`'s `None` return value to an exception, it probably makes more sense just to have `_egg_info_matches()` raise an exception in the first place.

Reply:
Or `_egg_info_matches` can be eliminated since this is the only place it is called :p

Reply:
Well, you have it as a well-tested, isolated function, so I think it helps to keep it broken out as a separate function. If you merged it, it would increase the complexity of the new function, and it would be harder to keep track of all the test cases and what they're probing (as there are already a number of them).